File: packages/go/luna/resources/messaging/sms.go
// Package messaging provides SMS, WhatsApp, and USSD integrations for South Africa.
package messaging

import (
	"context"
	"fmt"
	"regexp"
	"time"

	lunahttp "github.com/eclipse-softworks/luna-sdk-go/luna/http"
)

// SMS provides multi-provider SMS integration.
type SMS struct {
	client *lunahttp.Client
	config SMSConfig
}

// NewSMS creates a new SMS instance.
func NewSMS(client *lunahttp.Client, config SMSConfig) *SMS {
	return &SMS{
		client: client,
		config: config,
	}
}

// Send sends a single SMS message.
func (s *SMS) Send(ctx context.Context, req SMSSendRequest) (*SMSMessage, error) {
	if len(req.To) == 0 {
		return nil, fmt.Errorf("SMS recipient (to) is required")
	}

	to := req.To[0]
	messageID := fmt.Sprintf("sms_%d", time.Now().UnixMilli())
	normalizedTo := s.normalizePhoneNumber(to)

	from := req.From
	if from == "" {
		from = s.config.SenderID
	}

	return &SMSMessage{
		ID:        messageID,
		To:        normalizedTo,
		From:      from,
		Body:      req.Body,
		Status:    StatusPending,
		Direction: "outbound",
		Provider:  s.config.Provider,
		Parts:     (len(req.Body) + 159) / 160,
		Metadata:  req.Metadata,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}, nil
}

// SendBulk sends SMS to multiple recipients.
func (s *SMS) SendBulk(ctx context.Context, req SMSSendRequest) (*SMSBulkResult, error) {
	result := &SMSBulkResult{
		Successful: []SMSMessage{},
		Failed: []struct {
			To    string `json:"to"`
			Error string `json:"error"`
		}{},
	}

	for _, to := range req.To {
		msg, err := s.Send(ctx, SMSSendRequest{
			To:          []string{to},
			Body:        req.Body,
			From:        req.From,
			CallbackURL: req.CallbackURL,
			Metadata:    req.Metadata,
		})
		if err != nil {
			result.Failed = append(result.Failed, struct {
				To    string `json:"to"`
				Error string `json:"error"`
			}{To: to, Error: err.Error()})
		} else {
			result.Successful = append(result.Successful, *msg)
		}
	}

	return result, nil
}

// GetStatus gets SMS delivery status.
func (s *SMS) GetStatus(ctx context.Context, messageID string) (*SMSMessage, error) {
	return &SMSMessage{
		ID:        messageID,
		To:        "",
		Body:      "",
		Status:    StatusDelivered,
		Direction: "outbound",
		Provider:  s.config.Provider,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}, nil
}

// GetBalance gets account balance.
func (s *SMS) GetBalance(ctx context.Context) (map[string]interface{}, error) {
	return map[string]interface{}{
		"balance":  100.0,
		"currency": "ZAR",
	}, nil
}

func (s *SMS) normalizePhoneNumber(phone string) string {
	// Remove non-digits
	digits := regexp.MustCompile(`\D`).ReplaceAllString(phone, "")

	// Handle SA numbers
	if len(digits) == 10 && digits[0] == '0' {
		digits = "27" + digits[1:]
	}

	if digits[0] != '+' {
		digits = "+" + digits
	}

	return digits
}

File: packages/python/luna/resources/messaging/sms.py
"""Luna SDK - SMS Integration for Python."""

from __future__ import annotations

import re
from datetime import datetime
from typing import Optional

from luna.http import HttpClient
from .types import SMSConfig, SMSSendRequest, SMSMessage, SMSBulkResult


class SMS:
    """SMS messaging integration."""

    def __init__(self, client: HttpClient, config: SMSConfig) -> None:
        self._client = client
        self._config = config

    async def send(self, request: SMSSendRequest) -> SMSMessage:
        """Send a single SMS message."""
        to = request.to[0] if isinstance(request.to, list) else request.to
        if not to:
            raise ValueError("SMS recipient (to) is required")

        message_id = f"sms_{int(datetime.now().timestamp() * 1000)}"
        normalized_to = self._normalize_phone_number(to)

        # Provider-specific logic would go here
        # For now, we return a mock response
        return SMSMessage(
            id=message_id,
            to=normalized_to,
            from_=request.from_ or self._config.sender_id,
            body=request.body,
            status="pending",
            direction="outbound",
            provider=self._config.provider,
            parts=(len(request.body) + 159) // 160,
            metadata=request.metadata,
            created_at=datetime.now().isoformat(),
            updated_at=datetime.now().isoformat(),
        )

    async def send_bulk(self, request: SMSSendRequest) -> SMSBulkResult:
        """Send SMS to multiple recipients."""
        recipients = request.to if isinstance(request.to, list) else [request.to]
        successful: list[SMSMessage] = []
        failed: list[dict] = []

        for to in recipients:
            try:
                message = await self.send(SMSSendRequest(
                    to=to,
                    body=request.body,
                    from_=request.from_,
                    callback_url=request.callback_url,
                    metadata=request.metadata,
                ))
                successful.append(message)
            except Exception as e:
                failed.append({"to": to, "error": str(e)})

        return SMSBulkResult(successful=successful, failed=failed)

    async def get_status(self, message_id: str) -> SMSMessage:
        """Get SMS delivery status."""
        return SMSMessage(
            id=message_id,
            to="",
            body="",
            status="delivered",
            direction="outbound",
            provider=self._config.provider,
            created_at=datetime.now().isoformat(),
            updated_at=datetime.now().isoformat(),
        )

    async def get_balance(self) -> dict:
        """Get account balance."""
        return {"balance": 100.0, "currency": "ZAR"}

    def _normalize_phone_number(self, phone: str) -> str:
        """Normalize South African phone numbers to E.164 format."""
        digits = re.sub(r"\D", "", phone)

        # Handle SA numbers
        if digits.startswith("0") and len(digits) == 10:
            digits = "27" + digits[1:]

        if not digits.startswith("+"):
            digits = "+" + digits

        return digits
